File: Tree/2897. Apply Operations on Array to Maximize Sum of Squares.py
class Solution:
    def maxSum(self, nums: List[int], k: int) -> int:
        count = [0] * 30
        for a in nums:
            for i in range(30):
                if a & (1 << i):
                    count[i] += 1
        res = 0
        mod = 10 ** 9 + 7
        for j in range(k):
            cur = 0
            for i in range(30):
                if count[i]:
                    count[i] -= 1
                    cur |= 1 << i
            res = (res + cur * cur % mod) % mod
        return res

class Solution:
    def maxSum(self, nums: List[int], k: int) -> int:
        count = Counter(i for a in nums for i in range(32) if a & (1 << i))
        return sum(sum((count[i] > j) << i for i in range(32)) ** 2 for j in range(k)) % (10 ** 9 + 7)



class Solution:
    def maxSum(self, nums: List[int], k: int) -> int:
        n = len(nums)
        IDX = [0] * 32
        ans = [0] * n
        
        for x in nums:
            for i in range(32):
                if (x >> i) & 1:
                    ans[IDX[i]] += 1 << i
                    IDX[i] += 1
    
        res = 0 
        mod = 10**9 + 7
        
        for i in range(k):
            res += ans[i] ** 2
            res %= mod 
            
        return res

File: Tree/2897. Apply Operations on Array to Maximize Sum of Squares.cpp
/*
So apply operation as many as possible,
it will sort the array on each bit,
from random array [1,0,1,0,1,0,0],
to a sorted array [0,0,0,0,1,1,1].
*/

class Solution {
public:
    int maxSum(vector<int>& nums, int k) {
        vector<int>cnt(30);
        const long mod = 1e9 + 7;
        for(int i = 0; i<30; ++i){
            for(auto x: nums){
                if((1 << i) & x){
                    ++cnt[i];
                }
            }
        }
        long ans = 0;
        for(int i = 0; i<k; ++i){
            long cur = 0;
            for(int j = 0; j<30; ++j){
                if(cnt[j]-->0){
                    cur = cur | 1<<j;
                }
            }
            ans = (ans + cur * cur % mod) % mod;
        }
        return ans;        
    }
};


class Solution {
public:
    int maxSum(vector<int>& nums, int k) {
        int n = nums.size();
        vector<int> cnt(31, 0);
        
        for (int bit = 0; bit < 31; bit ++) {
            for (int j = 0; j < n; j ++) 
                if (nums[j] & (1 << bit)) cnt[bit] ++;
        }
        
        ll result = 0;
        
        for (int i = 0; i < k; i ++) {
            ll val = 0;
            
            for (int bit = 0; bit < 31; bit ++) {
                if (cnt[bit] < (i+1)) continue;
                val |= (1 << bit);
            }
            result = (result + val*val) % MOD;
        }
        return result;
    }
};